File: src/main/java/br/com/brforgers/ducts/Ducts.java
package br.com.brforgers.ducts;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import br.com.brforgers.ducts.blockentities.DuctBlockEntity;
import br.com.brforgers.ducts.blocks.DuctBlock;
import br.com.brforgers.ducts.inventories.DuctInventory;
import br.com.brforgers.ducts.screens.DuctScreen;
import net.minecraft.client.gui.screens.MenuScreens;
import net.minecraft.world.inventory.MenuType;
import net.minecraft.world.item.BlockItem;
import net.minecraft.world.item.CreativeModeTab;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.common.extensions.IForgeMenuType;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.fml.DistExecutor;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.ForgeRegistries;
import net.minecraftforge.registries.RegistryObject;

@Mod(Ducts.MOD_ID)
public class Ducts 
{
	public static final String MOD_ID = "ducts";
	public static final Logger LOGGER = LogManager.getLogger(Ducts.class);
	
	private static final DeferredRegister<Block> BLOCK_REGISTRY = DeferredRegister.create(ForgeRegistries.BLOCKS, MOD_ID);
	private static final DeferredRegister<Item> ITEM_REGISTRY = DeferredRegister.create(ForgeRegistries.ITEMS, MOD_ID);
	private static final DeferredRegister<BlockEntityType<?>> BLOCKENTITY_REGISTRY = DeferredRegister.create(ForgeRegistries.BLOCK_ENTITY_TYPES, MOD_ID);
	private static final DeferredRegister<MenuType<?>> MENUTYPE_REGISTRY = DeferredRegister.create(ForgeRegistries.MENU_TYPES, MOD_ID);

	public static final RegistryObject<Block> DUCT_BLOCK = BLOCK_REGISTRY.register("duct", () -> new DuctBlock());
	public static final RegistryObject<Item> DUCT_ITEM = ITEM_REGISTRY.register("duct", () -> new BlockItem(DUCT_BLOCK.get(), new Item.Properties().tab(CreativeModeTab.TAB_REDSTONE)));
	public static final RegistryObject<BlockEntityType<DuctBlockEntity>> DUCT_BLOCKENTITY_TYPE = BLOCKENTITY_REGISTRY.register("duct", () -> BlockEntityType.Builder.of(DuctBlockEntity::new, DUCT_BLOCK.get()).build(null));
	
	public static final RegistryObject<MenuType<DuctInventory>> DUCT_MENU = MENUTYPE_REGISTRY.register("duct", () -> IForgeMenuType.create(DuctInventory::new));
	
	public Ducts() 
	{
		final IEventBus modEventBus = FMLJavaModLoadingContext.get().getModEventBus();
		
		BLOCK_REGISTRY.register(modEventBus);
		ITEM_REGISTRY.register(modEventBus);
		BLOCKENTITY_REGISTRY.register(modEventBus);
		MENUTYPE_REGISTRY.register(modEventBus);
		
		DistExecutor.unsafeRunWhenOn(Dist.CLIENT, () -> () -> modEventBus.addListener(Ducts::onClientSetupEvent));
	}
	
	public static void onClientSetupEvent(FMLClientSetupEvent event) 
	{
		MenuScreens.register(DUCT_MENU.get(), DuctScreen::new);
	}
}

File: src/main/kotlin/br/com/brforgers/mods/ducts/Ducts.kt
package br.com.brforgers.mods.ducts

import br.com.brforgers.mods.ducts.blockentities.DuctBlockEntity
import br.com.brforgers.mods.ducts.blocks.DuctBlock
import br.com.brforgers.mods.ducts.inventories.DuctInventory
import br.com.brforgers.mods.ducts.screens.DuctScreen
import net.minecraft.block.Block
import net.minecraft.client.gui.ScreenManager
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.inventory.container.ContainerType
import net.minecraft.item.BlockItem
import net.minecraft.item.Item
import net.minecraft.item.ItemGroup
import net.minecraft.network.PacketBuffer
import net.minecraftforge.common.extensions.IForgeContainerType
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent
import net.minecraftforge.registries.ForgeRegistries
import org.apache.logging.log4j.Level
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import thedarkcolour.kotlinforforge.forge.KDeferredRegister
import thedarkcolour.kotlinforforge.forge.MOD_BUS
import thedarkcolour.kotlinforforge.forge.runForDist

@Mod(Ducts.ID)
object  Ducts {
    const val ID = "ducts"

    val LOGGER: Logger = LogManager.getLogger(ID)

    val DUCT_BLOCK: Block = DuctBlock()

    var DUCT_MENU: ContainerType<DuctInventory> = IForgeContainerType.create { id: Int, inv: PlayerInventory?, data: PacketBuffer ->
        DuctInventory(id, inv!!, data.readBlockPos())
    }

    init {
        LOGGER.log(Level.INFO, "Ducts!")

        val BLOCK_REGISTRY = KDeferredRegister(ForgeRegistries.BLOCKS, ID)
        BLOCK_REGISTRY.registerObject("duct") { DUCT_BLOCK }
        BLOCK_REGISTRY.register(MOD_BUS)

        val ITEM_REGISTRY = KDeferredRegister(ForgeRegistries.ITEMS, ID)
        ITEM_REGISTRY.registerObject("duct") { BlockItem(DUCT_BLOCK, Item.Properties().tab(ItemGroup.TAB_REDSTONE)) }
        ITEM_REGISTRY.register(MOD_BUS)

        val BLOCK_ENTITY_REGISTRY = KDeferredRegister(ForgeRegistries.TILE_ENTITIES, ID)
        BLOCK_ENTITY_REGISTRY.registerObject("duct") { DuctBlockEntity.type }
        BLOCK_ENTITY_REGISTRY.register(MOD_BUS)

        val CONTAINER_REGISTRY = KDeferredRegister(ForgeRegistries.CONTAINERS, ID)
        CONTAINER_REGISTRY.registerObject("duct") { DUCT_MENU }
        CONTAINER_REGISTRY.register(MOD_BUS)

        runForDist(
            clientTarget = {
                MOD_BUS.addListener(::onClientSetup)
            },
            serverTarget = {

            }
        )

    }

    private fun onClientSetup(event: FMLClientSetupEvent) {
        ScreenManager.register(DUCT_MENU, ::DuctScreen)
    }
}